gcsfs/core.py

```diff
@@ -1161,24 +1161,27 @@ def find(self, path, withdirs=False, detail=False, prefix="", **kwargs):
         sdirs = set()
         cache_entries = {}
         for o in out:
-            par = self._parent(o["name"])
-            if par not in sdirs:
-                sdirs.add(par)
-                dirs.append(
-                    {
-                        "Key": self.split_path(par)[1],
-                        "Size": 0,
-                        "name": par,
-                        "StorageClass": "DIRECTORY",
-                        "type": "directory",
-                        "size": 0,
-                    }
-                )
-            # Don't cache "folder-like" objects (ex: "Create Folder" in GCS console) to prevent
-            # masking subfiles in subsequent requests.
-            if not o["name"].endswith("/"):
-                cache_entries.setdefault(par, [])
-                cache_entries[par].append(o)
+            par = o["name"]
+            while par:
+                par = self._parent(par)
+                if par not in sdirs:
+                    if len(par) < len(path):
```

Comment:
I found the cause of this bug, the issue is that self._parent, doesn't work correctly for files with an empty filename (not really a surprise).

Example:
```
{'kind': 'storage#object',
 'id': 'bucket/something/items//1605786947866168',
 'selfLink': 'https://www.googleapis.com/storage/v1/b/bucket/something%2Fitems%2F',
 'mediaLink': 'https://www.googleapis.com/download/storage/v1/b/abucket/something%2Fitems%2F?generation=1605786947866168&alt=media',
 'name': 'bucket/something/items/',
 'bucket': 'bucket',
 'generation': '1605786947866168',
 'metageneration': '1',
 'contentType': 'application/octet-stream',
 'storageClass': 'STANDARD',
 'size': 0,
...
 'type': 'file'}
 ```

The parent for this spark created "file" called `bucket/something/items/` is the directory `bucket/something/items/`, but `_parent` will return `bucket/something/` instead

Reply:
Oh, you mean because there is an actual item whose name ends with "/"?

Reply:
cc https://github.com/dask/gcsfs/pull/355#pullrequestreview-606488402

Reply:
Sort of. the filename is actually `''` which is, if you prefix it with the bucket and directory: `'bucket/something/items/'` (which looks like a directory)

Reply:
I also just realized `ls` is not returning subdirectories anymore (`0.7.1` actually does include those)

Reply:
? Sorry, ca you please show this again?

Reply:
If I compare that with the builtins of fsspec, ls should actually include the directories as well.

Reply:
Yes yes, it should - and I think it does! We do have tests for `ls`, but if you come up with something showing the problem, I'll fix that too.

Reply:
That's perfectly fine with me :)
OK to merge from your view?

Reply:
@martindurant It's not fixing the actual issue. Also, I just figured out the other bug.

Reply:
I think this has to do with caching, let me create a test case

Reply:
This probably deserve's it's own bugreport. But let me include it here, as I think it is related to this code.
I first created the structure:
```
fs.touch('bucket/testcase/df/_SUCCESS')
fs.touch('bucket/testcase/df/')
fs.touch('bucket/testcase/df/sub/file')
```

Now I started a new python process (not entirely sure if this is needed):
```
>>> fs.ls('antenna-dataproc-test/testcase/df/')
['antenna-dataproc-test/testcase/df/',
 'antenna-dataproc-test/testcase/df/_SUCCESS',
 'antenna-dataproc-test/testcase/df/sub']
>>> fs.find('antenna-dataproc-test/testcase/df/')
['antenna-dataproc-test/testcase/df/',
 'antenna-dataproc-test/testcase/df/_SUCCESS',
 'antenna-dataproc-test/testcase/df/sub/file']
>>> fs.ls('antenna-dataproc-test/testcase/df/')
['antenna-dataproc-test/testcase/df/_SUCCESS']
```

Reply:
This is with 0.7.2 as it is on pypi.